packages/flutter/lib/src/widgets/scrollable_list.dart

```diff
@@ -111,15 +110,148 @@ class ListViewport extends VirtualViewport {
+  ScrollableLazyList({
+    Key key,
+    double initialScrollOffset,
+    Axis scrollDirection: Axis.vertical,
+    ScrollListener onScroll,
+    SnapOffsetCallback snapOffsetCallback,
+    double snapAlignmentOffset: 0.0,
+    this.itemExtent,
+    this.itemCount,
+    this.itemBuilder,
+    this.padding,
+    this.scrollableListPainter
+  }) : super(
+    key: key,
+    initialScrollOffset: initialScrollOffset,
+    scrollDirection: scrollDirection,
+    onScroll: onScroll,
+    snapOffsetCallback: snapOffsetCallback,
+    snapAlignmentOffset: snapAlignmentOffset
+  ) {
+    assert(itemExtent != null);
+    assert(itemBuilder != null);
+  }
+
+  final double itemExtent;
+  final int itemCount;
+  final ItemListBuilder itemBuilder;
+  final EdgeDims padding;
+  final ScrollableListPainter scrollableListPainter;
+
+  ScrollableState createState() => new _ScrollableLazyListState();
+}
+
+class _ScrollableLazyListState extends ScrollableState<ScrollableLazyList> {
+  ScrollBehavior createScrollBehavior() => new OverscrollBehavior();
+  ExtentScrollBehavior get scrollBehavior => super.scrollBehavior;
+
+  void _handleExtentsChanged(double contentExtent, double containerExtent) {
+    config.scrollableListPainter?.contentExtent = contentExtent;
+    setState(() {
+      scrollTo(scrollBehavior.updateExtents(
+        contentExtent: contentExtent,
+        containerExtent: containerExtent,
+        scrollOffset: scrollOffset
+      ));
+    });
+  }
+
+  void dispatchOnScrollStart() {
+    super.dispatchOnScrollStart();
+    config.scrollableListPainter?.scrollStarted();
+  }
+
+  void dispatchOnScroll() {
+    super.dispatchOnScroll();
+    config.scrollableListPainter?.scrollOffset = scrollOffset;
+  }
+
+  void dispatchOnScrollEnd() {
+    super.dispatchOnScrollEnd();
+    config.scrollableListPainter?.scrollEnded();
+  }
+
+  Widget buildContent(BuildContext context) {
+    return new LazyListViewport(
+      onExtentsChanged: _handleExtentsChanged,
+      startOffset: scrollOffset,
+      scrollDirection: config.scrollDirection,
+      itemExtent: config.itemExtent,
+      itemCount: config.itemCount,
+      itemBuilder: config.itemBuilder,
+      padding: config.padding,
+      overlayPainter: config.scrollableListPainter
+    );
+  }
+}
+
+class LazyListViewport extends _VirtualListViewport with VirtualViewportLazyMixin {
+  LazyListViewport({
+    Key key,
+    ExtentsChangedCallback onExtentsChanged,
+    double startOffset: 0.0,
+    Axis scrollDirection: Axis.vertical,
+    double itemExtent,
+    EdgeDims padding,
+    Painter overlayPainter,
+    this.itemCount,
+    this.itemBuilder
+  }) : super(
+    key,
+    onExtentsChanged,
+    startOffset,
+    scrollDirection,
+    itemExtent,
+    false, // Don't support wrapping yet.
+    padding,
+    overlayPainter
+  );
+
+  final int itemCount;
+  final ItemListBuilder itemBuilder;
+}
+
+class _ListViewportElement extends VirtualViewportElement<_VirtualListViewport> {
```

Comment:
This class should be moved up to after _VirtualListViewport, and renamed _VirtualListViewportElement.